### PR TITLE
Implement "WorldObject/Role" part of methods

### PR DESCRIPTION
**Pull request by a8901053**
# Changes
Implement "WorldObject/Role" turnTo,move,useWeapon and switchCurrentWeapon methods.

# Screenshot (Optional)

### CODE REVIEW

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/WorldObject.java [hunk 1]
```diff
@@ -4,10 +4,35 @@
...
 public abstract class WorldObject {
-    private Rectangle body;
-    private World world;
-    private float angle;
+    protected Rectangle body = new Rectangle(0, 0, 0, 0);
+    protected World world;
+    protected float angle;
     private int id;
 
+    public WorldObject() {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/WorldObject.java:14]: `WorldObject` can default its location, but its body's size should be explicitly passed from the constructor.
As the following, FYI:
```java
public WorldObject(Dimension size) {
    this.body =  new Rectangle(size);
}
```

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 2]
```diff
@@ -9,16 +9,14 @@
 import java.util.List;
 
 public class Player extends Role implements ClientApi {
+    private final static int WEAPON_SLOTS_AMOUNT = 2;
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:12]: `WEAPON_SLOTS_SIZE`

You should keep your naming **consistent**!!!!!!

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 3]
```diff
@@ -9,16 +9,14 @@
 import java.util.List;
 
 public class Player extends Role implements ClientApi {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:11]:
```java
/**
 * @author - a89010531111@gmail.com
 */
```

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java [hunk 4]
```diff
@@ -6,64 +6,109 @@
...
     private int maxHp;
-    private List<WeaponSlot> weaponSlots;
+    private final WeaponSlot[] weaponSlots;
     private List<Magazine> magazines;
     private WeaponSlot currentWeaponSlot;
 
-    public void turn(float angle) {
+    public Role(int weaponSlotsSize) {
+        weaponSlots = new WeaponSlot[weaponSlotsSize];
+        initWeaponSlots();
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java:27]: Great to extract it to another method. 👍

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java [hunk 5]
```diff
@@ -6,64 +6,109 @@
...
 
+    private void initWeaponSlots() {
+        for (int i = 0; i < weaponSlots.length; i++) {
+            weaponSlots[i] = new WeaponSlot();
+        }
     }
 
-    public void moveBackward() {
 
+    public void turnTo(float angle) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java:37]: `turnAngleTo`

We should give it a better name when it comes up to our brain.

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java [hunk 6]
```diff
@@ -6,64 +6,109 @@
...
     }
 
-    public void switchWeapon(int weaponIndex) {
+    public void pickUpWeapon() {
 
     }
 
     public void injure(Role attackRole, int damage) {
-
+        hp -= damage;
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java:79]: No `def` involved? You can create another issue for this :).

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java [hunk 7]
```diff
@@ -6,64 +6,109 @@
...
     }
 
     public boolean isDead() {
-        return false;
+        return hp <= 0;
     }
 
-    public void setCurrentWeapon(Weapon weapon) {
+    public void addWeaponToWeaponSlot(Weapon weapon, int weaponSlotIndex) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java:86]: rename it to `putWeaponToSlot`

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java [hunk 8]
```diff
@@ -6,64 +6,109 @@
...
+        weaponSlots[weaponSlotIndex].setWeapon(weapon);
+    }
 
+    public Weapon getWeaponByWeaponSlotIndex(int weaponSlotIndex) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Role.java:90]: `getWeaponBySlotIndex`

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/weapon/Gun.java [hunk 9]
```diff
@@ -1,11 +1,12 @@
...
 
+import java.util.List;
+
 public class Gun extends Weapon {
     private int bulletsShotAtOnce;
     private Magazine magazine;
 
-    public void reload(Magazine magazine) {
-
+    public void reload(List<Magazine> magazines) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/weapon/Gun.java:9]: Not sure why a list of magazines is passed here? 
Shouldn't the gun only be reloaded with one magazine?

File: server/src/test/java/tw/waterball/survive/io/server/worldobject/role/RoleTest.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+import org.junit.Test;
+import org.mockito.Mockito;
+import tw.waterball.survive.io.api.api.ClientApi;
+import tw.waterball.survive.io.server.worldobject.weapon.Gun;
+import tw.waterball.survive.io.server.worldobject.weapon.Weapon;
+
+/**
+ * @author - a89010531111@gmail.com
+ */
+
```
Comment (Johnny850807) [inline server/src/test/java/tw/waterball/survive/io/server/worldobject/role/RoleTest.java:13]: No Space.

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Thank you! Review done.

**Review by ShawnSWu — COMMENTED** [review]
LGTM

File: api/src/main/java/tw/waterball/survive/io/api/api/ServerApi.java [hunk 11]
```diff
@@ -8,21 +8,21 @@
 
     void reload(ReloadRequest reloadRequest);
 
-    void turn(TurnRequest turnRequest);
+    void turnTo(TurnToRequest turnToRequest);
```
[the remark this answers is not in the document]
Comment (a8901053) [reply api/src/main/java/tw/waterball/survive/io/api/api/ServerApi.java:11]: > Would it be better if it was name turnDirection()?
This is turn angle not direction.

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/weapon/Gun.java (hunk 9 above)
Comment (a8901053) [reply server/src/main/java/tw/waterball/survive/io/server/worldobject/weapon/Gun.java:9]: 如果傳入單個彈夾，也許會有裝到沒滿的彈夾，所以傳入玩家身上的所有彈夾去遍歷，讓很多彈夾的狀態下裝子彈一定能裝滿

**Review by Johnny850807 — APPROVED** [review]
ＬＧＴＭ

**Pull request merged**